File: tests/gpu/war_weariness_test.py
"""War weariness is scored PER BATTLE — the
tests/cpu/seats/war-weariness.test.ts twin.

    python tests/gpu/war_weariness_test.py

The scripted parity gate proves the two engines AGREE, never that either agrees
with Civ 6, and it reaches only the wars its seeds happen to fight — Ancient and
Classical bases, no city-state combat, no simultaneous wars. Every rule below is
poked directly instead, through the same seat-generic entry points both engines
call.

    WWP = (EraBase * Location) + Death

Location 1 at home / 2 abroad; Death = 3 * EraBase to the side that lost a unit;
any battle with a CITY at the abroad column; both sides score, "without any
discrimination".
"""

from __future__ import annotations

import sys
from pathlib import Path

import torch

sys.path.insert(0, str(Path(__file__).resolve().parent.parent.parent / "gpu"))
from core import BatchSim, load_rules, load_fixture, FIXTURES


def build():
    rules = load_rules()
    paths = sorted(FIXTURES.glob("seed*.json"))
    assert paths, "no fixtures — run `npm run seed && npm run export` first"
    sim = BatchSim([load_fixture(paths[0])], rules, device="cpu", dtype=torch.float64)
    assert sim.R >= 2, "needs two civs to poke simultaneous wars"
    for _ in range(20):
        sim.step()
    sim.war[:, 0, 1:1 + sim.R] = sim.war[:, 1:1 + sim.R, 0] = False
    sim.war[:, 1:1 + sim.R, 1:1 + sim.R] = False
    sim.sync_war()  # close the pokes under transpose
    sim.ww[:] = 0
    sim.ww_turn[:] = -1
    return sim


def one(sim) -> torch.Tensor:
    return torch.ones(sim.B, dtype=torch.bool)


def owned_tile(sim, seat: int) -> torch.Tensor:
    for t in range(sim.T):
        if bool((sim.tile_seat[:, t] == seat).all()):
            return torch.full((sim.B,), t, dtype=torch.long)
    raise AssertionError(f"no tile owned by seat {seat}")


def neutral_tile(sim) -> torch.Tensor:
    for t in range(sim.T):
        if bool((sim.tile_seat[:, t] < 0).all()):
            return torch.full((sim.B,), t, dtype=torch.long)
    raise AssertionError("no unowned tile")


def main() -> None:
    sim = build()
    rww = sim.rules.war_weariness
    abroad, death = int(rww["abroad"]), int(rww["death"])
    per = int(rww["perAmenity"])
    # Ancient SURPRISE — the one row of the sourced table where a casus belli
    # buys nothing (formal == surprise); the premium opens at Classical.
    b = int(rww["eraSurprise"][0])
    assert int(rww["eraFormal"][0]) == b, "Ancient formal and surprise are equal"
    assert int(rww["eraSurprise"][1]) > int(rww["eraFormal"][1]), "the premium opens at Classical"

    away, home = neutral_tile(sim), owned_tile(sim, 0)

    # --- both sides score, and the aggressor gets no discount --------------
    sim.ww[:] = 0
    sim._ww_battle(one(sim), 0, 1, away)
    assert int(sim.ww[0, 0, 1]) == b * abroad, int(sim.ww[0, 0, 1])
    assert int(sim.ww[0, 1, 0]) == b * abroad, int(sim.ww[0, 1, 0])
    print(f"  both sides score {b * abroad} on neutral ground")

    # --- HOME is half of ABROAD, per side, on the SAME battle --------------
    sim.ww[:] = 0
    sim._ww_battle(one(sim), 0, 1, home)
    assert int(sim.ww[0, 0, 1]) == b, "the defender at home pays a single base"
    assert int(sim.ww[0, 1, 0]) == b * abroad, "the invader pays double"
    print(f"  one battle, two multipliers: home {b} vs abroad {b * abroad}")

    # --- a CITY forces the abroad column for BOTH --------------------------
    sim.ww[:] = 0
    sim._ww_battle(one(sim), 0, 1, home, city=True)
    assert int(sim.ww[0, 0, 1]) == b * abroad, "a city drags its own defender abroad"
    print("  a city giving or receiving the attack scores at the abroad column")

    # --- a death costs the side that LOST the unit, and only that side -----
    sim.ww[:] = 0
    sim._ww_battle(one(sim), 0, 1, away, d_died=one(sim))
    assert int(sim.ww[0, 0, 1]) == b * abroad, "the survivor pays no death term"
    assert int(sim.ww[0, 1, 0]) == b * abroad + death * b, int(sim.ww[0, 1, 0])
    print(f"  the loser pays {death} more bases: {int(sim.ww[0, 1, 0])} vs {int(sim.ww[0, 0, 1])}")

    # --- BARBARIANS neither accrue it nor inflict it -----------------------
    sim.ww[:] = 0
    barb = int(sim.BARB_ROW)
    sim._ww_battle(one(sim), 0, barb, away, d_died=one(sim))
    sim._ww_battle(one(sim), barb, 0, away, d_died=one(sim))
    assert int(sim.ww.sum()) == 0, (
        "a barbarian fight scored war weariness. Every seat is permanently "
        "hostile to barbarians, so counting it makes 'at peace with everyone' "
        "unreachable for the whole game and no accumulator can ever drain"
    )
    print("  barbarians score nothing, in either direction")

    # --- a CITY-STATE is a real opponent but holds no accumulator ---------
    sim.ww[:] = 0
    citystate_row = 1 + sim.R
    sim._ww_battle(one(sim), 0, citystate_row, away, city=True)
    assert int(sim.ww[0, 0, citystate_row]) > 0, "warring a minor wears you down normally"
    assert int(sim.ww[0, citystate_row, :].sum()) == 0, "a minor keeps no accumulator"
    print("  a city-state is a valid opponent and holds nothing itself")

    # --- wars score SEPARATELY; only the worst is felt --------------------
    sim.ww[:] = 0
    sim._ww_battle(one(sim), 0, 1, away)
    sim._ww_battle(one(sim), 0, 1, away)
    sim._ww_battle(one(sim), 0, 2, away)
    step = b * abroad
    assert int(sim._ww_max(0)[0]) == step * 2, int(sim._ww_max(0)[0])
    assert int(sim._ww_sum(0)[0]) == step * 3, int(sim._ww_sum(0)[0])
    print(f"  the max {step * 2} is the worst war, NOT the sum {step * 3}")

    # --- decay: fought this turn / phoney / at peace with everyone --------
    sim.ww[:] = 0
    sim.ww_turn[:] = -1
    sim.ww[:, 0, 1] = 1000
    sim.ww[:, 0, 2] = 1000
    sim.ww_turn[:, 0, 1] = int(sim.turn)  # blood was spilled against civ 0
    sim.war[:, 0, 1 + 0] = sim.war[:, 1 + 0, 0] = True
    sim.sync_war()
    sim._ww_decay(0)
    assert int(sim.ww[0, 0, 1]) == 1000, "a war fought THIS turn does not decay"
    assert int(sim.ww[0, 0, 2]) == 1000 - int(rww["decayAtWar"]), int(sim.ww[0, 0, 2])
    sim.turn += 1
    sim.war[:, 0, 1 + 0] = sim.war[:, 1 + 0, 0] = False
    sim.sync_war()
    sim._ww_decay(0)
    assert int(sim.ww[0, 0, 1]) == 1000 - int(rww["decayAtPeace"]), int(sim.ww[0, 0, 1])
    print(f"  decay: 0 fought, {int(rww['decayAtWar'])} phoney, {int(rww['decayAtPeace'])} at peace")

    # --- a peace treaty settles THAT war and no other ---------------------
    sim.ww[:] = 0
    sim.ww[:, 0, 1] = 900
    sim.ww[:, 0, 2] = 900
    sim._ww_peace(one(sim), 0, 1)
    assert int(sim.ww[0, 0, 1]) == 0, "the treaty sheds 2000, floored at zero"
    assert int(sim.ww[0, 0, 2]) == 900, "the OTHER war is untouched"
    print(f"  a treaty sheds {int(rww['peaceTreaty'])} from one war only")

    # --- the amenity conversion, with no ceiling -------------------------
    sim.ww[:] = 0
    sim.ww[:, 0, 1] = per * 12 + (per - 1)
    assert int(sim._ww_penalty(0)[0]) == 12, int(sim._ww_penalty(0)[0])
    assert int(sim._ww_penalty(1)[0]) == 0, "civ 0 has fought nothing"
    sim.ww[:, 0, 1] = per - 1
    assert int(sim._ww_penalty(0)[0]) == 0, "the remainder buys nothing"
    print(f"  {per} points buy one amenity, remainder lost, and 12 is reachable (no cap)")

    # --- the accumulator round-trips snapshot/restore ---------------------
    sim.ww[:] = 0
    sim.ww[:, 0, 1] = 4321
    sim.ww_turn[:, 0, 1] = 7
    snap = sim.snapshot()
    sim.ww[:, 0, 1] = 0
    sim.ww_turn[:, 0, 1] = -1
    sim.restore(snap)
    assert int(sim.ww[0, 0, 1]) == 4321, "ww lost in snapshot/restore"
    assert int(sim.ww_turn[0, 0, 1]) == 7, "ww_turn lost in snapshot/restore"
    print("  ww and ww_turn round-trip snapshot/restore")

    # --- a war DECLARED but never fought costs nothing -------------------
    sim = build()
    sim.war[:, 0, 1 + 0] = sim.war[:, 1 + 0, 0] = True
    sim.sync_war()
    for _ in range(30):
        sim.step()
    assert int(sim._ww_max(0)[0]) == 0, (
        f"a phoney war accrued {int(sim._ww_max(0)[0])}. The per-BATTLE model's "
        "whole point is that DECLARING a war costs nothing — the old flat "
        "+1/turn could not tell a phoney war from a bloody one"
    )
    print("  30 turns of a declared, unfought war cost nothing")

    print("WAR WEARINESS OK — per-battle accrual, per-war accumulators, no ceiling")


if __name__ == "__main__":
    main()

File: tests/cpu/seats/war-weariness.test.ts
import { describe, it, expect } from 'vitest';
import { warWearinessBattle, warWearinessTurn, warWearinessPeace, wwGet, wwMax, wwSum, wwEraBase } from '../../../cpu/core/weariness';
import { WW_ERA_BASE_FORMAL, WW_ERA_BASE_SURPRISE, WW_ABROAD_MULT, WW_DEATH_MULT, WW_DECAY_AT_WAR, WW_DECAY_AT_PEACE, WW_PEACE_TREATY, WAR_WEARINESS_PER_AMENITY, warWearinessPenalty } from '../../../cpu/data/seats';
import { createGame } from '../../../cpu/core/game';
import { settleFirstCity } from '../helpers';
import { seatOf, setTileOwner, setWar, BARB_SEAT, seatOfCityState } from '../../../cpu/core/seats';
import type { GameState, Seat } from '../../../cpu/core/types';

// War weariness is scored PER BATTLE.
//
//     WWP = (EraBase * Location) + Death
//
// Source: the Civ 6 wiki's War weariness page and its reference, CivFanatics
// thread 623207. The SHAPE is what matters: a declared war that nobody fights
// costs nothing and drains away, while one bloody turn is expensive.
//
// Every assertion below is on the SEAT-GENERIC entry points. There is no
// per-seat path to test separately.

function newGame(opponents = 2): GameState {
  const state = createGame({
    width: 44, height: 26, seed: 4210,
    withResources: true, withWonders: false, unitsMode: true,
    withVillages: false, cityStates: 1, opponents,
  });
  settleFirstCity(state, 0);
  return state;
}

/** A tile owned by `seat`, and one owned by nobody. */
function tiles(state: GameState, seat: number): { home: number; away: number } {
  const home = state.map.tiles.find((t) => t.terrain !== 'OCEAN' && t.terrain !== 'COAST')!;
  setTileOwner(home, seat, -1);
  const away = state.map.tiles.find((t) => t.index !== home.index && t.terrain !== 'OCEAN')!;
  setTileOwner(away, -1, -1);
  return { home: home.index, away: away.index };
}

describe('#51/S7.8f war weariness — the per-battle model', () => {
  it('scores BOTH sides, at the base for the era, with no discount for the aggressor', () => {
    const state = newGame();
    const { away } = tiles(state, 0);
    const civ = 1;
    warWearinessBattle(state, 0, civ, away);
    // Ancient, no casus belli anywhere: 16, doubled for fighting off both
    // civs' land. "Accumulated by both the attacker and the defender, without
    // any discrimination."
    const base = WW_ERA_BASE_SURPRISE[0];
    expect(wwGet(seatOf(state, 0)!, civ)).toBe(base * WW_ABROAD_MULT);
    expect(wwGet(seatOf(state, civ)!, 0)).toBe(base * WW_ABROAD_MULT);
  });

  it('fighting at HOME is half what fighting abroad is — per side, on the same battle', () => {
    const state = newGame();
    const { home } = tiles(state, 0); // SEAT 0's borders
    const civ = 1;
    warWearinessBattle(state, 0, civ, home);
    const base = WW_ERA_BASE_SURPRISE[0];
    // the defender is at home, the invader is not — one battle, two multipliers
    expect(wwGet(seatOf(state, 0)!, civ)).toBe(base);
    expect(wwGet(seatOf(state, civ)!, 0)).toBe(base * WW_ABROAD_MULT);
  });

  it('a CITY giving or receiving the attack forces the abroad column for both', () => {
    const state = newGame();
    const { home } = tiles(state, 0);
    const civ = 1;
    warWearinessBattle(state, 0, civ, home, { city: true });
    // same tile as the test above, where seat 0 scored a single base
    expect(wwGet(seatOf(state, 0)!, civ)).toBe(WW_ERA_BASE_SURPRISE[0] * WW_ABROAD_MULT);
  });

  it('a death costs the side that LOST the unit three more bases, and only that side', () => {
    const state = newGame();
    const { away } = tiles(state, 0);
    const civ = 1;
    warWearinessBattle(state, 0, civ, away, { dDied: true });
    const base = WW_ERA_BASE_SURPRISE[0];
    expect(wwGet(seatOf(state, 0)!, civ)).toBe(base * WW_ABROAD_MULT);
    expect(wwGet(seatOf(state, civ)!, 0)).toBe(base * WW_ABROAD_MULT + WW_DEATH_MULT * base);
  });

  it('the era table is the sourced one, and the surprise premium opens at Classical', () => {
    const state = newGame();
    const p = seatOf(state, 0)!;
    expect(wwEraBase(state, 0, 1)).toBe(WW_ERA_BASE_SURPRISE[0]);
    // Ancient is the one row where a casus belli buys nothing — 16 either way.
    expect(WW_ERA_BASE_FORMAL[0]).toBe(WW_ERA_BASE_SURPRISE[0]);
    for (let e = 1; e < WW_ERA_BASE_FORMAL.length; e++) {
      expect(WW_ERA_BASE_SURPRISE[e]).toBeGreaterThan(WW_ERA_BASE_FORMAL[e]);
    }
    // the seat's OWN era drives it: a Classical civ wearies at the Classical row
    p.research.techs = ['MINING', 'BRONZE_WORKING', 'CURRENCY', 'WRITING', 'ASTROLOGY', 'HORSEBACK_RIDING', 'IRON_WORKING'];
    const era = wwEraBase(state, 0, 1);
    expect(WW_ERA_BASE_SURPRISE as readonly number[]).toContain(era);
  });

  it('BARBARIANS neither accrue it nor inflict it', () => {
    const state = newGame();
    const { away } = tiles(state, 0);
    warWearinessBattle(state, 0, BARB_SEAT, away, { dDied: true });
    warWearinessBattle(state, BARB_SEAT, 0, away, { dDied: true });
    expect(wwSum(seatOf(state, 0)!)).toBe(0);
    // ...which is what keeps "at peace with everyone" reachable at all: every
    // seat is permanently hostile to barbarians.
    expect(wwSum(seatOf(state, BARB_SEAT))).toBe(0);
  });

  it('a CITY-STATE is a real opponent but keeps no accumulator of its own', () => {
    const state = newGame();
    const { away } = tiles(state, 0);
    const cityState = seatOfCityState(state.cityStates![0].id);
    warWearinessBattle(state, 0, cityState, away, { city: true });
    expect(wwGet(seatOf(state, 0)!, cityState)).toBeGreaterThan(0);
    expect(wwSum(seatOf(state, cityState))).toBe(0);
  });

  it('wars score SEPARATELY and only the worst is felt', () => {
    const state = newGame(2);
    const { away } = tiles(state, 0);
    const a = 1;
    const b = 2;
    warWearinessBattle(state, 0, a, away);
    warWearinessBattle(state, 0, a, away);
    warWearinessBattle(state, 0, b, away);
    const one = WW_ERA_BASE_SURPRISE[0] * WW_ABROAD_MULT;
    expect(wwGet(seatOf(state, 0)!, a)).toBe(one * 2);
    expect(wwGet(seatOf(state, 0)!, b)).toBe(one);
    expect(wwMax(seatOf(state, 0)!)).toBe(one * 2); // NOT the sum
    expect(wwSum(seatOf(state, 0)!)).toBe(one * 3);
  });

  it('a war fought this turn does not decay; a phoney war sheds 50 and peace sheds 200', () => {
    const state = newGame(2);
    const { away } = tiles(state, 0);
    const a = 1;
    const b = 2;
    const p = seatOf(state, 0)!;
    p.ww = { [a]: 1000, [b]: 1000 };
    p.wwTurn = { [a]: state.turn }; // blood was spilled against `a` this turn
    setWar(state, (state.seats[a] as Seat).seat, 0, true); // seat 0 is at war with somebody
    warWearinessTurn(state, 0);
    expect(wwGet(p, a)).toBe(1000);
    expect(wwGet(p, b)).toBe(1000 - WW_DECAY_AT_WAR);

    // ...and with every war over, the drain is four times faster. The turn has
    // to move on for the `a` war to count as unfought — the stamp is what tells
    // a war being fought from a phoney one.
    state.turn += 1;
    setWar(state, (state.seats[a] as Seat).seat, 0, false);
    warWearinessTurn(state, 0);
    expect(wwGet(p, a)).toBe(1000 - WW_DECAY_AT_PEACE);
    expect(wwGet(p, b)).toBe(1000 - WW_DECAY_AT_WAR - WW_DECAY_AT_PEACE);
    expect(away).toBeGreaterThanOrEqual(0);
  });

  it('a peace treaty sheds 2000 from THAT war only, and never below zero', () => {
    const state = newGame(2);
    const a = 1;
    const b = 2;
    const p = seatOf(state, 0)!;
    p.ww = { [a]: 900, [b]: 900 };
    warWearinessPeace(state, 0, a);
    expect(wwGet(p, a)).toBe(0); // 900 - 2000, floored
    expect(wwGet(p, b)).toBe(900);
    expect(WW_PEACE_TREATY).toBeGreaterThan(WW_DECAY_AT_PEACE);
  });

  it('400 points buy one amenity and the remainder buys nothing', () => {
    expect(warWearinessPenalty(0)).toBe(0);
    expect(warWearinessPenalty(WAR_WEARINESS_PER_AMENITY - 1)).toBe(0);
    expect(warWearinessPenalty(WAR_WEARINESS_PER_AMENITY)).toBe(1);
    expect(warWearinessPenalty(WAR_WEARINESS_PER_AMENITY * 3 + 399)).toBe(3);
    // deleted the ceiling: a long, bloody war keeps costing.
    expect(warWearinessPenalty(WAR_WEARINESS_PER_AMENITY * 12)).toBe(12);
  });

  it('a war DECLARED but not fought costs nothing', () => {
    const state = newGame();
    const civ = 1;
    setWar(state, 0, civ, true);
    for (let i = 0; i < 30; i++) {
      state.turn += 1;
      warWearinessTurn(state, 0);
      warWearinessTurn(state, civ);
    }
    expect(wwMax(seatOf(state, 0)!)).toBe(0);
    expect(wwMax(seatOf(state, civ))).toBe(0);
  });
});
